### PR TITLE
Query-side token replacement

### LINKED ISSUE #54: Add query-time mechanism for applying token replacements
When we load structures back up again, we’ll need to load their token replacements and then apply them to incoming lookups. Steps:

- [ ] in the FuzzyPhraseSet constructor, it will load its metadata which will supply a string→string mapping of things to replace. Once we have our prefix set loaded, we’ll want to turn each string (the from side and the to side) into IDs, by looking them all up, and build an ID→ID mapping to use going forward (probably a `HashMap<u32,u32>`).
- [ ] in each place where we do lookups, after we’ve looked up the ID for each word (either in the prefix graph or the fuzzy graph), before we construct our QueryWord, we’ll look up the ID we’ve just retrieved in our in-memory hashmap, and if we find it, we’ll swap out the ID we were going to use for its replacement, then proceed as usual
- [ ] TBD: we’ll need to figure out what to do with the last word if we’re doing a prefix match — if the word we’re maybe replacing is both an entire, replaceable word and the prefix of other words, we’ll probably want to generate two output candidates, a FullWord with edit distance 0 with the replaced ID, and the prefix range as-is (no replacement) and test them both; there will probably be changes needed to the prefix graph to get it to supply the necessary metadata about what it found, to communicate whether it’s only a prefix match, only an exact match, or both, to decide what to do here
- [ ] functions to modify:
  - [ ] contains / contains_prefix
  - [ ] get_terminal_word_possibilities / get_nonterminal_word_possibilities (these wrap the lookup logic for fuzzy_match, fuzzy_match_prefix, fuzzy_match_windows, and fuzzy_match_multi)

### PR DESCRIPTION
**Pull request by apendleton**
~**Note:** this PR continues https://github.com/mapbox/fuzzy-phrase/pull/62 ; if that one is reviewed/merged first, this one will be easier to read.~ EDIT: merged

This PR makes a bunch of changes to facilitate doing token replacement at query time (cf. the index-time half of the work in #58). To wit:
- [x] ~**PrefixSet refactor:** the work in #62, which makes the PrefixSet lookup functions more expressive~ EDIT: merged upstream
- [x] **loading logic for replacements:** take the data that was serialized in #58 and parse it at structure load time (fixing errors related to the structure initially failing to load); the process here is to take each word's string representation and map it to a word ID, and then build an ID-to-ID map to facilitate query time lookups; we use a BTreeMap here to allow for range lookups as well as exact lookups, to make prefix matching against token replacement candidates efficient
- [x] **enhancements to `get_nonterminal_word_possibilities`**, which is the inner function used by fuzzy_match for all words, fuzzy_match_prefix for all words except the last one, and fuzzy_match_windows and fuzzy_match_multi in comparable contexts to figure out all the word IDs that might be intended by a word in a given position; this function now does 1-to-1 token replacement after lookup and, if applicable, spelling correction
- [x] **enhancements to `get_terminal_word_possibilities`**, which is the inner function used by fuzzy_match_prefix for final words, and fuzzy_match_windows and fuzzy_match_multi in comparable contexts, to figure out all the word IDs and/or prefix ranges that might be intended by a word in the last position of a phrase; this function now does some new stuff:
    - [x] if a word is spelling-corrected, consider token replacement as per the nonterminal case
    - [x] if it's prefix-matched, check if it's the beginning of one or more token replacements, and if so, emit a full word for each replacement (e.g., if it's the "stre" prefix, emit the "st" full word); note that we only emit the full word if it's not included in the proposed prefix, so for the word "s", even though we have a "street" => "s" replacement, we don't emit "st" because we'd match "st" anyway by scanning for "s"
    - [x] if replacement words are emitted but there are additional terminations in the prefix, emit the prefix as well (e.g., if the user types "for" and a replacement exists for "fort" => "ft" but the lexicon also includes the word "fortenberry", emit both an "ft" full word and a "for" prefix); alternatively, if all prospective completions are token replacements, emit only the full word, and if no prospective completions are token replacements, emit only the range
    - [x] when considering emitting a range, if the word prefix lands on a final position in the graph (that is to say, represents a whole word), and that word doesn't have any completions (that is to say, it's not the prefix of any additional longer words), emit a full word instead of a prefix; this isn't connected to token replacement, but is a slight optimization/semantic improvement in preparation for #56
- [x] **addition of index-time replacement:** the above item ended up being a sufficiently powerful construction to solve #57 ahead of schedule, so we now don't have a reason to double-index phrases with both pre- and post-replacement forms for purposes of autocomplete; as such, we can go ahead and apply the replacements at index time
- [x] **enhancement of `contains` and `contains_prefix`:** we don't actually use `contains` or `contains_prefix` in carmen but it was silly to not have them support token replacement when everything else did, so they now do too; note that this required essentially rewriting contains_prefix because the correct replacement-aware prefix-matching logic means there might be more than one possibility in the terminal position, so a straight linear search is no longer practical.
- [x] **lots of tests:** there are a pile of new tests in `glue` to test the interaction of these various pieces of new functionality. I put them in their own file to keep the main file from getting prohibitively long (and might want to similarly move the remaining tests out)
    - [x] note that I moved the `load_word_replacements` test introduced in #58 into this new file for consistency, and shortened it to just reuse the existing replacement-containing structure I was already building for all the other tests
- [x] **benchmarks:** added a glue benchmark variant that uses the English geocoder-abbrevations replacements

fixes #54, #55, #57

### CODE REVIEW

**Comment by apendleton** [comment]
Hrm, looks like this doesn't build with Rust 1.26.2 because it uses inclusive ranges. Rewriting to avoid them would be a pain... maybe now's the time to move us to Travis?

**Comment by apendleton** [comment]
**UPDATE:** moved to travis in #65

File: src/glue/mod.rs [hunk 1]
```diff
@@ -238,7 +246,8 @@ impl FuzzyPhraseSet {
 
         let metadata_reader = BufReader::new(fs::File::open(directory.join(Path::new("metadata.json")))?);
         let metadata: FuzzyPhraseSetMetadata = serde_json::from_reader(metadata_reader)?;
-        if metadata != FuzzyPhraseSetMetadata::default() {
+        let default = FuzzyPhraseSetMetadata::default();
+        if metadata.index_type != default.index_type || metadata.format_version != default.format_version {
```
Comment (boblannon) [inline src/glue/mod.rs]: Just checking for comprehension: This check is to make sure that the serialized metadata is in step with the type and version that the code expects.  If we made a significant change in the future, we'd bump the version number in the code, making the old metadata incompatible with the new code.

File: src/glue/mod.rs [hunk 2]
```diff
@@ -365,23 +412,57 @@ impl FuzzyPhraseSet {
...
+                    if target < found_range.0 || target > found_range.1 {
+                        Some(target)
+                    } else {
+                        None
+                    }
+                ).collect();
+
+            // if everything within our range will get token-replaced, don't emit the unreplaced word
+            if num_terminations != replacements.len() {
+                // if there's just one word and it's final, emit a full_word
```
Comment (boblannon) [inline src/glue/mod.rs]: the comments here were a little hard to follow, so i'm writing them out to make sure i get it.  

If `num_terminations != replacements.len()`, then that means that at least some of the terminations will not be replaced, and so we push the range onto `last_variants`. Later on, we avoid adding anything from `replacements` to `last_variants` if they already exist in `last_variants`.

but, if `num_terminations == replacements.len()` then that would mean that all of the possible terminations are going to be replaced.  we'll leave `last_variants` empty and just add all of the replacements to it later.

This is sortof implementing `last_variants` as a Set, except that the check for membership of a new value only looks at `QueryWord::Full` elements?

File: src/glue/mod.rs (hunk 1 above)
Comment (apendleton) [reply src/glue/mod.rs]: Correct. The reason for the change is that until we added token replacements to the metadata file, the metadata file was character-for-character identical for each index generated, and I was lazy before and just checked to make sure it was exactly the same as the default. But now it might vary, because of token replacements, so the check is more specific.

File: src/glue/mod.rs [hunk 3]
```diff
@@ -310,24 +337,34 @@ impl FuzzyPhraseSet {
...
                 None => { return Ok(false) }
             }
         }
-        Ok(self.phrase_set.contains_prefix(QueryPhrase::new(&id_phrase)?)?)
+        match self.get_terminal_word_possibilities(phrase[last_idx].as_ref(), 0)? {
+            Some(possibilities) => word_possibilities.push(possibilities),
+            None => return Ok(false),
+        }
+
+        let phrase_matches = self.phrase_set.match_combinations_as_prefixes(&word_possibilities, 0)?;
```
Comment (boblannon) [inline src/glue/mod.rs]: it's neat that you could re-use `match_combinations_as_prefixes` here

File: src/glue/mod.rs (hunk 2 above)
Comment (apendleton) [reply src/glue/mod.rs]: So at this point we've already emitted full words for each word in our range that's a token replacement. If that's all of the things in our range, we're done -- we don't have to emit any more. So like, if the word is "str" and that completes to "street" which gets replaced by "st", we'll have emitted a full word for "st". If that's the entirety of the range, we don't have to also emit a range for "str" -- it won't match anything anyway.

If that's *not* the case (which will be the norm, as most words won't be token-replaced), we do have to emit something. If our word is the prefix for a bunch of words, we'll emit a range. If our word is a whole word and nothing follows it, we can just emit a full word, which makes for a slightly more efficient phrase graph scan. So like, if the word is "massachusetts", we found that word, and it's a whole word, and it's not the beginning of any other words, so we rather than emitting a QueryWord::Prefix (which would signify "things that start with 'massachusetts'"), we can just emit 'massachusetts' as a full word instead.

File: src/glue/replacement_tests.rs [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+            TEST_SET.fuzzy_match_prefix(&["100", "main", "s"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "main", "st"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "main", "str"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "main", "stre"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "main", "stree"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "main", "street"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "fort", "wayne", "road"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "ft", "wayne", "road"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "fort", "wayne", "rd"], 1, 1).unwrap(),
+            TEST_SET.fuzzy_match_prefix(&["100", "ft", "wayne", "rd"], 1, 1).unwrap(),
```
Comment (boblannon) [inline src/glue/replacement_tests.rs]: this is almost ASCII art

File: src/glue/mod.rs [hunk 5]
```diff
@@ -678,7 +759,7 @@ impl FuzzyPhraseSet {
         };
 
         // fuzzy-lookup all the words, but only once apiece (per prefix-y-ness type)
-        let mut all_words: HashMap<(&str, bool), Vec<QueryWord>> = HashMap::new();
+        let mut all_words: FxHashMap<(&str, bool), Vec<QueryWord>> = FxHashMap::default();
```
Comment (boblannon) [inline src/glue/mod.rs]: based on the `rustc-hash` docs it seems like this was chosen as a speed boost.  any other reason?

File: src/glue/mod.rs (hunk 2 above)
Comment (boblannon) [reply src/glue/mod.rs]: So I understood what was meant to happen just from reading the PR description. What i was trying to get more concrete about was what was happening line-by-line.  The inline comments just threw me off a bit, like: 

```rust
            // if everything within our range will get token-replaced, don't emit the unreplaced word
            if num_terminations != replacements.len() {
```

The comment and the next line of code both start with an "if" and so you sortof expect them to be describing the same condition, but they're describing opposite ones.  They're not contradictory, but it feels that way at first.

File: Cargo.toml [hunk 6]
```diff
@@ -14,6 +14,7 @@ serde_json = "1.0"
 rmp-serde = "0.13.7"
```
Comment (boblannon) [inline Cargo.toml]: won't let me comment on an earlier line, but it's probably about time to bump the version number, right? ~especially because node-fuzzy-phrase will need to change?~ scratch that, maybe not! It looks like all it requires is a vector of objects, and uses `neon_serde` to convert them.

your call.

File: src/glue/mod.rs (hunk 2 above)
Comment (apendleton) [reply src/glue/mod.rs]: They're describing unrelated conditions. Each comment describes the thing being tested in the if-statement immediately beneath it. The first relates to token replacements, but the second does not, it relates to whether or not there are continuations of the current potentially-partial word. We just only consider this condition if we've cleared the first hurdle. Regardless, let me know how you think I could word it to be more clear.

File: src/glue/mod.rs (hunk 5 above)
Comment (apendleton) [reply src/glue/mod.rs]: Nope, just speed.

File: Cargo.toml (hunk 6 above)
Comment (apendleton) [reply Cargo.toml]: Ah yeah, I can bump the version number, but yeah, it doesn't affect node-fuzzy-phrase, which we just point at this repo by gitsha.

**Comment by apendleton** [comment]
Going to go ahead and merge this, but can incorporate any feedback into #66
